pkg/networkservice/common/connect/client.go

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2020 Cisco and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connect
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/serialize"
+)
+
+type connectClient struct {
+	connection *networkservice.Connection
+	cancel     context.CancelFunc
+	executor   serialize.Executor
+}
+
+// NewClient - client chain element for use with single incoming connection, translates from incoming server connection to
+//             outgoing client connection
+func NewClient(cancel context.CancelFunc) networkservice.NetworkServiceClient {
+	return &connectClient{
+		cancel: cancel,
+	}
+}
+
+func (c *connectClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
+	var conn *networkservice.Connection
+	var err error
+	<-c.executor.AsyncExec(func() {
```

Comment:
Can we get rid of using `executor` here? Because we are not using async here and look like we just creating additional goroutine and just join.
```suggestion
```

Reply:
You are correct, mutex is the better choice here as we simply need exclusivity.

I had somehow convinced myself we also needed to preserve order (which executor does and mutex does not)... but as I tried to write the explanation of why, I found I was mistaken :)